### DISCUSSION TITLE
Optional Multiple File Uploads Broken When Using Pydantic v2.3.0

### DISCUSSION THREAD

**Discussion by jbourquin** [fastapi/fastapi#10280] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, UploadFile, File
from typing import Annotated, List

app = FastAPI()

@app.post("/")
async def upload_attachments(
    attachment_files: Annotated[List[UploadFile] | None, File(alias='attachments', description='Multiple attachments')] = []
):
  file_list = []
  if attachment_files:
    for f in attachment_files:
      file_list.append(f.filename)
  
  return {"file_list": file_list}
```

### Description

FastAPI (v0.103.1) raises a list type error when multiple file uploads are specified as optional and Pydantic v2.3.0 is installed. With the Pydantic v2.3.0, the example code returns the following response:

> Note: I am using HTTPie to generate the request but the error is the same for other clients like Postman

```bash
http -f POST localhost:8000 attachments@./document.pdf attachments@./image.jpg
```

```json
{
    "detail": [
        {
            "input": {
                "file": {
                    "_TemporaryFileArgs": {
                        "buffering": -1,
                        "dir": null,
                        "encoding": null,
                        "errors": null,
                        "mode": "w+b",
                        "newline": null,
                        "prefix": null,
                        "suffix": null
                    },
                    "_file": {},
                    "_max_size": 1048576,
                    "_rolled": false
                },
                "filename": "image.jpg",
                "headers": {
                    "content-disposition": "form-data; name=\"attachments\"; filename=\"image.jpg\"",
                    "content-type": "image/jpeg"
                },
                "size": 57694
            },
            "loc": [
                "body",
                "attachments"
            ],
            "msg": "Input should be a valid list",
            "type": "list_type",
            "url": "https://errors.pydantic.dev/2.3/v/list_type"
        }
    ]
}
```

However, downgrading to Pydantic v1.10.12 returns the correct response:

```bash
http -f POST localhost:8000 attachments@./document.pdf attachments@./image.jpg
```

```json
{
    "file_list": [
        "document.pdf",
        "image.jpg"
    ]
}
```

### Operating System

macOS

### FastAPI Version

0.103.1

### Pydantic Version

2.3.0

### Python Version

3.10.9

**Comment by pawel-kmiecik** [comment]
Looks like it's still bugged - confirmed multiple file uploads does not work with pydantic2

**Comment by JavierSanchezCastro** [comment]
Hi!

This works for me:

```python3
from fastapi import FastAPI, UploadFile

app = FastAPI()

@app.post("/")
async def upload_attachments(attachment_files: list[UploadFile] = []):
  file_list = []
  print(attachment_files)
  if attachment_files:
    for f in attachment_files:
      file_list.append(f.filename)
  
  return {"file_list": file_list}
```

``fastapi==0.109.0
pydantic==2.5.2``

**Reply by JavierSanchezCastro** [reply]
I just see some endpoints of my project are failing with Annotated typing. I'm going to check in which version fails

**Reply by jbourquin** [reply]
The bug is related to the multiple uploads as being optional, i.e. you can either pass a list of file objects or None. So in my use case I have form that has the option of attaching multiple files, but they are not required.

**Comment by progre55** [comment]
Still seems to be broken on `pydantic==2.7`

**Comment by pedroconceicao** [comment]
I think my problem is related to this discussion so I'm sharing it here:

I have an API that has an endpoint that accepts a variable amount of files. The endpoint is defined in the following way:
```
@router_my_router.post('/stream', response_model=PredictionResponseModel)
async def my_endpoint(image_files: Optional[List[UploadFile]] = File([], alias='imageFiles')):
    return my_endpoint.predict(image_files)
```

Using `fastapi <= 0.104.0` the swagger looks like
![image](https://github.com/tiangolo/fastapi/assets/32837064/dabcf5cf-6055-4213-a5e4-cc6a494a373c)

however, using `fastapi > 0.104.0` the file input field is broken
![image](https://github.com/tiangolo/fastapi/assets/32837064/b59dec95-8493-4bf7-a387-054f2a8b8fd5)

not allowing me to upload files and make a request.

In both cases the versions of `pydantic` and `pydantic_core` are `2.7.1` and `2.18.2`, respectively. I'm not sure however if this is a `fastapi` issue 
or an issue of one of its dependencies.

**Comment by msukmanowsky** [comment]
Despite the fact that these types look completely wrong, this does seem to work (using `pydantic==2.6.4, fastapi==0.111.0, starlette==0.37.2`):

```python
@app.post(
    "/upload_files",
)
async def upload_files(
    files: list[UploadFile | str] = File(None),
):
    for file in files or []:
        print("file: %r %r", file.filename, file.content_type)
        contents = await file.read()
        print("contents: %r", contents[:100])
    return {"result": "ok"}
```

Tested for:

1. No files at all (`files` is `None`)
2. One file (`files` is a 1-length list)
3. Multiple files (`files` is an n-length list)

Would still appreciate someone having a look at this as it'd be nice to do:

```python
files: Optional[list[UploadFile]] = None
```

**Reply by ArtemBalandin81** [reply]
Hi! 
Unfortunately it doesn't work for me with suggested dependencies. 
Any combinations besides `list[UploadFile]` crash the ability to upload files in swagger.
And `list[UploadFile] = File(None)` raise `"Value error, Expected UploadFile, received: <class 'str'>"`
if there is no file to upload.

**Reply by msukmanowsky** [reply]
I recall reading somewhere that Swagger has a known issue with file uploads, you may want to try CURL/Postman/Python directly.

**Comment by obonyojimmy** [comment]
The trick is to **NOT** check the ***send empty value***  checkbox , whwn testing with swagger UI (see screenshot below). Tested with 

```
python = "^3.10"
fastapi = "^0.112.1"
pydantic = "^2.8.2"
```

The endpoint defination: 

```
@router.post("/messages")
async def send_chat_message(
	attachments: Union[List[UploadFile], None] = None,
) -> Message:
	print(attachments)
```

![image](https://github.com/user-attachments/assets/263e5d4e-29a8-4ca2-9817-5e1fee8bb4e3)

Tested for:

- No files at all (files is None)
- One file (files is a 1-length list)
- Multiple files (files is an n-length list)

**Comment by siddharthmanu** [comment]
I was facing this same issue today. Found the solution was simply to update FastAPI >= 0.113.0

Source: https://github.com/fastapi/fastapi/discussions/11131

**Comment by falcolnic** [comment]
some updates?

**Comment by sanjeev-bhandari** [comment]
Any Update on this issue

Similar issue in fastapi==0.135.3

Code:
```python
@router.post("", response_model=FeedbackRead)
async def create_feedback(
    ...
    screenshots: Annotated[
        list[UploadFile], File(description="Upload screenshot images")
    ] = [],
    .....
):
```

File upload shows showing noisy string instead of file Upload filed
<img width="627" height="196" alt="image" src="https://github.com/user-attachments/assets/c2a28302-7a85-40fd-a08f-9611ee9ae360" />

**Reply by sanjeev-bhandari** [reply]
Found workaround this:

https://github.com/fastapi/fastapi/discussions/14975#discussioncomment-15903138
